fdbclient/NativeAPI.actor.cpp

```diff
@@ -580,38 +580,44 @@ ACTOR static Future<Void> monitorClientInfo( Reference<AsyncVar<Optional<Cluster
 
 			choose {
 				when( ClientDBInfo ni = wait( clusterInterface->get().present() ? brokenPromiseToNever( clusterInterface->get().get().openDatabase.getReply( req ) ) : Never() ) ) {
-					TraceEvent("ClientInfoChange").detail("ChangeID", ni.id);
 					outInfo->set(ni);
 
-					if (ni.proxies.empty()) {
-						TraceEvent("ClientInfo_NoProxiesReturned").detail("ChangeID", ni.id);
-						continue;
-					} else if (!FlowTransport::transport().isClient()) {
-						continue;
-					}
-
-					vector<Future<Void>> onProxyFailureVec;
-					bool skipWaitForProxyFail = false;
-					for (const auto& proxy : ni.proxies) {
-						if (proxy.provisional) {
-							skipWaitForProxyFail = true;
+					loop {
+						TraceEvent("ClientInfoChange").detail("ChangeID", outInfo->get().id);
+						if (outInfo->get().proxies.empty()) {
+							TraceEvent("ClientInfo_NoProxiesReturned").detail("ChangeID", outInfo->get().id);
+							break;
+						} else if (!FlowTransport::transport().isClient()) {
 							break;
 						}
 
-						onProxyFailureVec.push_back(
+						state vector<Future<Void>> onProxyFailureVec;
+						bool skipWaitForProxyFail = false;
+						for (const auto& proxy : outInfo->get().proxies) {
+							if (proxy.provisional) {
+								skipWaitForProxyFail = true;
+								break;
+							}
+
+							onProxyFailureVec.push_back(
 						    IFailureMonitor::failureMonitor().onStateEqual(
 						        proxy.getConsistentReadVersion.getEndpoint(), FailureStatus()) ||
 						    IFailureMonitor::failureMonitor().onStateEqual(proxy.commit.getEndpoint(), FailureStatus()) ||
 						    IFailureMonitor::failureMonitor().onStateEqual(
 						        proxy.getKeyServersLocations.getEndpoint(), FailureStatus()) ||
 						    IFailureMonitor::failureMonitor().onStateEqual(
 						        proxy.getStorageServerRejoinInfo.getEndpoint(), FailureStatus()));
-					}
-					if (skipWaitForProxyFail) continue;
+						}
+						if (skipWaitForProxyFail) break;
 
-					leaderMon = Void();
-					wait(waitForAny(onProxyFailureVec));
-					leaderMon = ccf ? monitorLeader(ccf, clusterInterface) : Void();
+						leaderMon = Void();
+						state Future<Void> anyFailures = waitForAny(onProxyFailureVec);
+						wait(anyFailures || outInfo->onChange());
```

Comment:
Earlier, this was the only place `outInfo` could be changed, but since that's not true maybe we should add a comment mentioned that `outInfo` can be changed if a proxy responds with forwarding info.